exporters/fluentd/src/fluentd_exporter.cc

```diff
@@ -0,0 +1,435 @@
+ * @brief Create FluentD exporter with default options
+*/
+FluentdExporter::FluentdExporter() : options_(FluentdExporterOptions())
+{
+  Initialize();
+}
+
+/**
+ * @brief Create new Recordable
+ * @return Recordable
+*/
+std::unique_ptr<sdk::trace::Recordable> FluentdExporter::MakeRecordable() noexcept
+{
+  return std::unique_ptr<sdk::trace::Recordable>(new Recordable);
+}
+
+/**
+ * @brief Export spans.
+ * @param spans 
+ * @return Export result.
+*/
+sdk::common::ExportResult FluentdExporter::Export(
+    const nostd::span<std::unique_ptr<sdk::trace::Recordable>> &spans) noexcept
+{
+  // Calculate number of batches
+  seq_batch_++;
+
+  // If no spans in Recordable, then return error.
+  if (spans.size() == 0)
+  {
+    return sdk::common::ExportResult::kFailure;
+  }
+
+  json events = {};
+  {
+    // Write all Spans first
+    json obj = json::array();
+    obj.push_back(FLUENT_VALUE_SPAN);
+    json spanevents = json::array();
+    for (auto &recordable : spans)
+    {
+      auto rec = std::unique_ptr<Recordable>(static_cast<Recordable *>(recordable.release()));
+      if (rec != nullptr)
+      {
+        auto span = rec->span();
+        // Emit "Span" as fluentd event
+        json record = json::array();
+        record.push_back(span["options"][FLUENT_FIELD_ENDTTIME]);
+        json fields = {};
+        for (auto &kv : span["options"].items())
+        {
+          fields[kv.key()] = kv.value();
+        }
+        record.push_back(fields);
+        spanevents.push_back(record);
+        // Gather Span event statistics
+        seq_span_++;
+        LOG_TRACE("exporting: batch #%d span #%d", seq_batch_, seq_span_);
+        // Iterate over all events added on this span
+        for (auto &v : span["events"])
+        {
+          auto &event      = v[1];
+          std::string name = event[FLUENT_FIELD_NAME];
+
+          event[FLUENT_FIELD_SPAN_ID]  = span["options"][FLUENT_FIELD_SPAN_ID];
+          event[FLUENT_FIELD_TRACE_ID] = span["options"][FLUENT_FIELD_TRACE_ID];
+
+          // Event Time flows as a separate field in fluentd.
+          // However, if we may need to consider addding
+          // span["options"][FLUENT_FIELD_TIME]
+          // 
+          // Complete list of all Span attributes :
+          // for (auto &kv : span["options"].items()) { ... }
+
+          // Group all events by matching event name in array.
+          // This array is translated into FluentD forward payload.
+          if (!events.contains(name))
+          {
+            events[name] = json::array();
+          }
+          events[name].push_back(v);
+          // Gather event-on-Span statistics
+          seq_evt_++;
+          LOG_TRACE("exporting: batch #%d evt #%d", seq_batch_, seq_evt_);
+        }
+      }
+    }
+    obj.push_back(spanevents);
+    LOG_TRACE("sending %zu Span event(s)", obj[1].size());
+    std::vector<uint8_t> msg = nlohmann::json::to_msgpack(obj);
+    // Schedule upload of Span event(s)
+    Enqueue(msg);
```

Comment:
So, we are batching the events here, and sending them through a separate thread, to optimize the number of TCP connections? This is not as per the specs, as the `Exporter::Export()` method should really be doing the export through the transmission channels, and the return status should be the actual status of transmission : https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#exportbatch

There is already batch span processor to handle any such batching needs, but shouldn't be here as part of exporter.